File: verifiers/v1/harnesses/codex/tool_hook.mjs
/** Bridge Codex's awaited native tool hooks to the rollout's /tool policy. */

import { readFileSync } from "node:fs";

const hook = JSON.parse(readFileSync(0, "utf8"));
const before = hook.hook_event_name === "PreToolUse";

async function run() {
  const url = process.env.VF_CODEX_TOOL_URL;
  const secret = process.env.VF_CODEX_TOOL_SECRET;
  if (!url || !secret) throw new Error("Tool interception is not configured");

  let content = "";
  if (!before) {
    content =
      typeof hook.tool_response === "string"
        ? hook.tool_response
        : JSON.stringify(hook.tool_response);
    if (!content) content = `(${hook.tool_name} completed with no output)`;
  }

  let resultPrefix = "";
  let resultSuffix = "";
  if (before) {
    const command = hook.tool_input?.command;
    if (
      (hook.tool_name === "Bash" || hook.tool_name === "apply_patch") &&
      typeof command === "string"
    ) {
      resultPrefix = "Command blocked by PreToolUse hook: ";
      resultSuffix = `. Command: ${command}`;
    } else {
      resultPrefix = "Tool call blocked by PreToolUse hook: ";
      resultSuffix = `. Tool: ${hook.tool_name}`;
    }
  }

  const response = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${secret}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      phase: before ? "before" : "after",
      content: "nonempty_text",
      resultPrefix,
      resultSuffix,
      message: {
        role: "tool",
        tool_call_id: hook.tool_use_id,
        content,
        name: hook.tool_name,
      },
    }),
    signal: AbortSignal.timeout(30_000),
  });
  if (!response.ok) throw new Error(`Tool interception returned ${response.status}`);
  const decision = await response.json();
  if (!["allow", "rewrite", "stop"].includes(decision.action)) {
    throw new Error("Tool interception returned an invalid action");
  }

  let replacement = content;
  if (decision.action === "rewrite") {
    replacement = decision.message?.content;
    if (typeof replacement !== "string" || !replacement) {
      throw new Error("Tool interception returned an invalid replacement");
    }
  } else if (decision.action === "stop") {
    replacement = decision.reason || "Rollout terminated by interception.";
  } else if (before) {
    return;
  }

  if (before) {
    console.log(
      JSON.stringify({
        hookSpecificOutput: {
          hookEventName: "PreToolUse",
          permissionDecision: "deny",
          permissionDecisionReason: replacement,
        },
      }),
    );
  } else {
    // Codex PostToolUse has no updatedToolOutput field. Its universal feedback
    // response replaces the model-visible result without blocking the agent loop.
    console.log(JSON.stringify({ continue: false, stopReason: replacement }));
  }
}

run().catch((error) => {
  const reason = `Tool interception is unavailable: ${error.message}`;
  console.log(
    JSON.stringify(
      before
        ? {
            hookSpecificOutput: {
              hookEventName: "PreToolUse",
              permissionDecision: "deny",
              permissionDecisionReason: reason,
            },
          }
        : { continue: false, stopReason: reason },
    ),
  );
});

File: verifiers/v1/harnesses/codex/tool_hook.py
# /// script
# requires-python = ">=3.11,<3.15"
# dependencies = ["httpx", "mcp>=1.24,<2"]
# ///
"""Expose the rollout tool policy as a model-hidden Codex hook tool."""

import json
import os
from pathlib import Path

from mcp.server.fastmcp import FastMCP

# {tool_interception}

credentials_path = Path(os.environ.pop("VF_TOOL_INTERCEPTION_CONFIG"))
credentials = json.loads(credentials_path.read_bytes())
credentials_path.unlink()
interceptor = ToolInterceptionClient(  # noqa: F821
    credentials["url"], credentials["secret"]
)
server = FastMCP("vf-interceptor")


def output(decision: dict) -> str:
    if decision["action"] == "allow":
        return "{}"
    replacement = decision.get("message", {}).get("content")
    reason = (
        replacement or decision.get("reason") or "Rollout terminated by interception."
    )
    if not isinstance(reason, str) or not reason:
        raise TypeError("Codex hooks require a non-empty text replacement")
    return json.dumps(
        {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "deny",
                "permissionDecisionReason": reason,
            }
        }
    )


@server.tool()
def before(tool_name: str, tool_use_id: str, tool_input: dict) -> str:
    body = {
        "phase": "before",
        "content": "nonempty_text",
        "detachedParent": "exec",
        "message": {
            "role": "tool",
            "tool_call_id": tool_use_id,
            "content": "",
            "name": tool_name,
        },
    }
    command = tool_input.get("command")
    if tool_name in {"Bash", "apply_patch"} and isinstance(command, str):
        body["rewrite_prefix"] = "Command blocked by PreToolUse hook: "
        body["rewrite_suffix"] = f". Command: {command}"
    else:
        body["rewrite_prefix"] = "Tool call blocked by PreToolUse hook: "
        body["rewrite_suffix"] = f". Tool: {tool_name}"
    decision = interceptor.request(body)
    return output(decision)


try:
    server.run(transport="stdio")
finally:
    interceptor.close()
